### ISSUE TITLE
 Early Stopping Round Parameter (stopping_rounds) in XGBoost is not working properly with PySparkling Water

### ISSUE THREAD

**Issue by Shiutang-Li** [h2oai/sparkling-water#1089]
**Description**
When building an XGBoost model with either cross validation or a single validation set, when the validation metric is not improved in the number of rounds determined by stopping_rounds, it should stop training. Thus the actual number of trees used in the model would be less than what's assigned before training (ntrees).

But based on the codes below, stopping_rounds is not doing its job properly. It just keeps training until it reaches the preassigned ntrees. Because more trees then expected are used to build the model, the model obviously overfits.

**Testing Version**
h2o-pysparkling-2.4   (2.4.6)   on Spark 2.4.0

**Testing Data and Python Codes to prepare it**
I used kaggle's home credit default risk data to do the test. Download application_train.csv from
https://www.kaggle.com/c/home-credit-default-risk/data

and this is the code I used to prepare train / test set:

import pandas as pd
import numpy as np
df = pd.read_csv('......./application_train.csv',
                 sep=',',  header = 0)
df = df[df.NAME_CONTRACT_TYPE == 'Cash loans']
features = [
 'AMT_INCOME_TOTAL',
 'AMT_CREDIT',
 'AMT_ANNUITY',
 'AMT_GOODS_PRICE',
 'DAYS_BIRTH',
 'DAYS_EMPLOYED',
 'DAYS_REGISTRATION',
 'DAYS_ID_PUBLISH', 
 'EXT_SOURCE_1',
 'EXT_SOURCE_2',
 'EXT_SOURCE_3',
 'HOUR_APPR_PROCESS_START',
 'REG_REGION_NOT_LIVE_REGION',
 'REG_REGION_NOT_WORK_REGION',
 'LIVE_REGION_NOT_WORK_REGION',
 'REG_CITY_NOT_LIVE_CITY',
 'REG_CITY_NOT_WORK_CITY',
 'LIVE_CITY_NOT_WORK_CITY'
]

df[:200000][['SK_ID_CURR','TARGET'] + features].to_csv( '.... HCDR_train.csv', index = False)
df[200000:][['SK_ID_CURR','TARGET'] + features].to_csv( '.... HCDR_test.csv', index = False)

**Testing Codes - 1 (for comparison purpose): Python XGBoost on a single machine**
import xgboost as xgb
train_DM = xgb.DMatrix(train[features],label=train['TARGET'])
test_DM  = xgb.DMatrix(test[features], label=test['TARGET'])
params = {
              'seed': 0,
              'eta':  0.1,
              'max_leaves' : 60,
              'colsample_bytree': 1,
              'subsample': 1,
              'lambda': 1,
              'gamma':  0,
              'tree_method' :'hist',
              'grow_policy' :'lossguide',
              'objective'   : 'binary:logistic',
              'eval_metric' : 'auc',
              'silent' : 1
          } 
cv_table = xgb.cv(params = params,
                  dtrain = train_DM,
                  num_boost_round = 2000,
                  nfold  = 5,
                  stratified = True, 
                  shuffle = True,
                  early_stopping_rounds= 10,
                  verbose_eval= 5,
                  seed = 0
                 )
print(cv_table.shape[0])       # 105

_Test AUC is around 0.74 with different seeds. Or, instead of doing cv, use watch list:_
watchlist = [(train_DM, 'train'), (test_DM, 'valid')]
model  = xgb.train(params=params,  
                   dtrain=train_DM, 
                   num_boost_round=2000,
                   evals=watchlist,
                   early_stopping_rounds=10,
                   verbose_eval=5
                  )   # Stopping. Best iteration:     [138]	train-auc:0.802294	valid-auc:0.75009

**Testing Codes - 2: XGBoost estimator in PySparkling**
train = h2o.import_file( "...HCDR_train.csv",  sep=',')
test = h2o.import_file("...HCDR_test.csv",  sep=',')
train['TARGET'] = train['TARGET'].asfactor()
test['TARGET']  = test['TARGET'].asfactor()
features = [
 'AMT_INCOME_TOTAL',
 'AMT_CREDIT',
 'AMT_ANNUITY',
 'AMT_GOODS_PRICE',
 'DAYS_BIRTH',
 'DAYS_EMPLOYED',
 'DAYS_REGISTRATION',
 'DAYS_ID_PUBLISH', 
 'EXT_SOURCE_1',
 'EXT_SOURCE_2',
 'EXT_SOURCE_3',
 'HOUR_APPR_PROCESS_START',
 'REG_REGION_NOT_LIVE_REGION',
 'REG_REGION_NOT_WORK_REGION',
 'LIVE_REGION_NOT_WORK_REGION',
 'REG_CITY_NOT_LIVE_CITY',
 'REG_CITY_NOT_WORK_CITY',
 'LIVE_CITY_NOT_WORK_CITY'
]
xgb_estimator = H2OXGBoostEstimator(
    ntrees = 2000,
    learn_rate = 0.1,
    max_leaves = 60,
    stopping_rounds = 10,
    stopping_metric = "AUC",
    tree_method="hist",
    grow_policy="lossguide",
    seed=0)
xgb_estimator.train(x = features, y = 'TARGET', training_frame = train, validation_frame=test)
print(xgb_estimator.params['ntrees'])    # {'actual': 2000, 'default': 50} <- should be much less than 2000 

print(xgb_estimator.model_performance(train[features + ['TARGET']]).auc())
print(xgb_estimator.model_performance(test[features + ['TARGET']]).auc())  # 0.8018227468242209,  0.6965668370478819  <- cuz too many trees were built, it overfits

_CV is not working properly either_:
cv_xgb = H2OXGBoostEstimator(
    ntrees = 2000,
    learn_rate = 0.1,
    max_leaves = 60,
    stopping_rounds = 10,
    stopping_metric = "AUC",
    tree_method="hist",
    grow_policy="lossguide",
    nfolds=5, 
    seed=0)
cv_xgb.train(x = features, y = 'TARGET', training_frame = train)
cv_xgb.params['ntrees']  # {'actual': 2000, 'default': 50} <- should be much less than 2000 

print(cv_xgb.model_performance(train[features + ['TARGET']]).auc())
print(cv_xgb.model_performance(test[features + ['TARGET']]).auc()) # 0.8018227468242209,  0.6965668370478819

**Comment by jborchma** [comment]
In order for the early stopping to work correctly, you need to specify `score_tree_interval=xxx` to some non-zero number. This should solve your issue.

**Comment by Shiutang-Li** [comment]
Thanks! I'll look into that argument and test it.

**Comment by Shiutang-Li** [comment]
With score_tree_interval = 1, I was able to generate results which are similar to the ones generated by XGB in Python. Appreciated.
